### DISCUSSION TITLE
avg_num_neighbors

### DISCUSSION THREAD

**Discussion by tueboesen** [mir-group/nequip#163] [Q&A] [answered]
I am wondering why the dataset average number of neighbors is used rather than just computing the actual average number of neighbors per frame? 
The reason why I ask is that I might want to use something similar to nequip on a protein dataset with vastly different number of particles and neighbors between each data point.

**Comment by simonbatzner** [comment]
That would not be smooth as atoms leave and enter the environment.

**Comment by tueboesen** [comment]
What do you mean with smooth when we are looking at snapshots?
In my case, I would go to the extreme and only have 1 snapshot of each configuration, so talking about atoms leaving and entering wouldn't really be applicable.

**Comment by simonbatzner** [comment]
I am speaking about inference. When you use the potential in a MD simulation, atoms will enter and leave the local environments. If you compute the actual local number of neighbors of an atom, this will lead to a non-smooth change. 

Maybe we are talking about different things: can you explain in detail your situation? (for terminology, we use the word "frame" for a DFT frame, i.e. a collection of N atoms with an energy and the word "snapshot" for a local frame, i.e. the environment around an atom)

**Comment by tueboesen** [comment]
I haven't entirely settled on how I'm solving this problem yet, I'm still exploring options.
But I'm working on proteins, and the idea right now is to have a neural network that can quickly predict the binding energy per residue or perhaps even per residue pair of a protein.
So the network would never be used to actually do MD simulations with and it would only see one frame of each protein. I might also include forces, but the primary idea would be to just predict energy.
The problem is of course if the predicted energies also ends up shifting due to the shift in average number of neighbors.

**Comment by Linux-cpp-lisp** [comment]
Is it actually true that the average number of neighbors differs _that_ much between your different dataframes?

Given the application you describe it is certainly possible to do this per frame, and would be a fairly simple modification to the code if you want to try it for yourself, but I struggle to see how that would help— the goal of the normalization here is just to keep the scale of the numbers at initialization from being too crazy (to support stable SGD); if you change it per frame aren't you hurting your model's ability to learn common structure across frames?

**Comment by simonbatzner** [comment, marked as answer]
Sounds great. You asked why we use the dataset average number of neighbors, hence my answer since we built a ML potential and there it's a good idea. That being said, you can of course make other choices. If you want to use local # neighbors, we're happy to accept a PR if you think in your case it helps. I would suggest trying the default setup first however, I don't believe it will be an obstacle.

**Comment by tueboesen** [comment]
Okay I think I understand its purpose. 
My understanding is: the average number of neighbors is set in order to try and keep the data values close to  order 1 during its propagation through the neural network, in order to prevent vanishing or exploding gradients. Is this correct?

If that is indeed the case I will do some test runs with my data and see whether it is sufficient with just the average value. If not I guess I will have to find other ways to stabilize it.

The actual differences in number of neighbors would be rather high I suspect, probably something where it goes from 5-50 between different samples.

**Comment by Linux-cpp-lisp** [comment]
> My understanding is: the average number of neighbors is set in order to try and keep the data values close to order 1 during its propagation through the neural network, in order to prevent vanishing or exploding gradients. Is this correct?

Yes, this is correct.

> If that is indeed the case I will do some test runs with my data and see whether it is sufficient with just the average value. If not I guess I will have to find other ways to stabilize it.

Sounds good— as Simon said, if the average is not sufficient but the per-frame average (computed by the model in the forward pass) _is_ successful, we'd be happy to work on a PR upstreaming this.

Thanks!
